Vaadin HotswapAgent with Spring Boot throws NPE

HotswapAgent Vaadin plugin does not work with Spring Boot. This is because reflection does not find the right method when the servlet class is `SpringServlet` rather than `VaadinServlet` (see first exception in server log below). Hence, an NPE is thrown on every live reload. Also, the push connection is not established due to `/vaadinServlet` mapping not handled properly by gizmo.

Versions: 
* Vaadin 15
* Spring starter with Flow 3.2.alpha2 and devtools disabled
* HotswapAgent 1.4.1-SNAPSHOT

Server log:
```
/Users/joheriks/devtools/dcevm-11.0.6+1/Contents/Home/bin/java -agentlib:jdwp=transport=dt_socket,address=127.0.0.1:51009,suspend=y,server=n -Dmaven.multiModuleProjectDirectory=/Users/joheriks "-Dmaven.home=/Applications/IntelliJ IDEA.app/Contents/plugins/maven/lib/maven3" "-Dclassworlds.conf=/Applications/IntelliJ IDEA.app/Contents/plugins/maven/lib/maven3/bin/m2.conf" "-Dmaven.ext.class.path=/Applications/IntelliJ IDEA.app/Contents/plugins/maven/lib/maven-event-listener.jar" -javaagent:/Users/joheriks/Library/Caches/JetBrains/IntelliJIdea2020.1/captureAgent/debugger-agent.jar -Dfile.encoding=UTF-8 -classpath "/Applications/IntelliJ IDEA.app/Contents/plugins/maven/lib/maven3/boot/plexus-classworlds-2.6.0.jar:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar" org.codehaus.classworlds.Launcher -Didea.version2020.1 -DskipTests=true spring-boot:run
Connected to the target VM, address: '127.0.0.1:51009', transport: 'socket'
HOTSWAP AGENT: 22:09:26.304 INFO (org.hotswap.agent.HotswapAgent) - Loading Hotswap agent {1.4.1-SNAPSHOT} - unlimited runtime class redefinition.
HOTSWAP AGENT: 22:09:26.817 INFO (org.hotswap.agent.config.PluginRegistry) - Discovered plugins: [Hotswapper, WatchResources, AnonymousClassPatch, ClassInitPlugin, JdkPlugin, Hibernate, Hibernate3JPA, Hibernate3, Spring, Jersey1, Jersey2, Jetty, Tomcat, ZK, Logback, Log4j2, MyFaces, Mojarra, Omnifaces, ELResolver, WildFlyELResolver, OsgiEquinox, Owb, Proxy, WebObjects, Weld, JBossModules, ResteasyRegistry, Deltaspike, GlassFish, Vaadin, Wicket, CxfJAXRS, FreeMarker, Undertow, MyBatis]
Starting HotswapAgent '/Users/joheriks/devtools/dcevm-11.0.6+1/Contents/Home/lib/hotswap/hotswap-agent.jar'
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------< com.example:spring-skeleton >---------------------
[INFO] Building Project base for Spring Boot and Vaadin Flow 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] >>> spring-boot-maven-plugin:2.2.0.RELEASE:run (default-cli) > test-compile @ spring-skeleton >>>
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:resources (default-resources) @ spring-skeleton ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] Copying 1 resource
[INFO] 
[INFO] --- vaadin-maven-plugin:15.0.5:prepare-frontend (default) @ spring-skeleton ---
[INFO] writing file /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/flow-frontend/package.json.
[INFO] Copying frontend resources from jar files ...
[INFO] Visited 150 resources. Took 78 ms.
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ spring-skeleton ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:testResources (default-testResources) @ spring-skeleton ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ spring-skeleton ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] <<< spring-boot-maven-plugin:2.2.0.RELEASE:run (default-cli) < test-compile @ spring-skeleton <<<
[INFO] 
[INFO] 
[INFO] --- spring-boot-maven-plugin:2.2.0.RELEASE:run (default-cli) @ spring-skeleton ---
[INFO] Attaching agents: []
HOTSWAP AGENT: 22:09:34.838 INFO (org.hotswap.agent.HotswapAgent) - Loading Hotswap agent {1.4.1-SNAPSHOT} - unlimited runtime class redefinition.
HOTSWAP AGENT: 22:09:35.114 INFO (org.hotswap.agent.config.PluginRegistry) - Discovered plugins: [Hotswapper, WatchResources, AnonymousClassPatch, ClassInitPlugin, JdkPlugin, Hibernate, Hibernate3JPA, Hibernate3, Spring, Jersey1, Jersey2, Jetty, Tomcat, ZK, Logback, Log4j2, MyFaces, Mojarra, Omnifaces, ELResolver, WildFlyELResolver, OsgiEquinox, Owb, Proxy, WebObjects, Weld, JBossModules, ResteasyRegistry, Deltaspike, GlassFish, Vaadin, Wicket, CxfJAXRS, FreeMarker, Undertow, MyBatis]
Starting HotswapAgent '/Users/joheriks/devtools/dcevm-11.0.6+1/Contents/Home/lib/hotswap/hotswap-agent.jar'

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.2.0.RELEASE)

HOTSWAP AGENT: 22:09:35.809 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.spring.SpringPlugin' initialized in ClassLoader 'jdk.internal.loader.ClassLoaders$AppClassLoader@67424e82'.
HOTSWAP AGENT: 22:09:35.809 INFO (org.hotswap.agent.plugin.spring.SpringPlugin) - Spring plugin initialized - Spring core version '5.2.0.RELEASE'
2020-04-29 22:09:35.851  INFO 24632 --- [           main] org.vaadin.example.Application           : Starting Application on pandion.local with PID 24632 (/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/classes started by joheriks in /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring)
2020-04-29 22:09:35.853  INFO 24632 --- [           main] org.vaadin.example.Application           : No active profile set, falling back to default profiles: default
HOTSWAP AGENT: 22:09:35.892 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hibernate.HibernatePlugin' initialized in ClassLoader 'jdk.internal.loader.ClassLoaders$AppClassLoader@67424e82'.
HOTSWAP AGENT: 22:09:35.987 INFO (org.hotswap.agent.plugin.spring.SpringPlugin) - Registering basePackage org.vaadin.example
HOTSWAP AGENT: 22:09:36.276 INFO (org.hotswap.agent.plugin.vaadin.VaadinPlugin) - Initialized Vaadin plugin
2020-04-29 22:09:36.696  INFO 24632 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
2020-04-29 22:09:36.703  INFO 24632 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2020-04-29 22:09:36.704  INFO 24632 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.27]
HOTSWAP AGENT: 22:09:36.931 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.tomcat.TomcatPlugin' initialized in ClassLoader 'TomcatEmbeddedWebappClassLoader
  context: ROOT
  delegate: true
----------> Parent Classloader:
jdk.internal.loader.ClassLoaders$AppClassLoader@67424e82
'.
HOTSWAP AGENT: 22:09:36.932 INFO (org.hotswap.agent.plugin.tomcat.TomcatPlugin) - Tomcat plugin initialized - Tomcat version '9.0.27.0'
2020-04-29 22:09:36.971  INFO 24632 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2020-04-29 22:09:36.971  INFO 24632 --- [           main] o.s.web.context.ContextLoader            : Root WebApplicationContext: initialization completed in 1082 ms
HOTSWAP AGENT: 22:09:37.021 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.vaadin.VaadinPlugin' initialized in ClassLoader 'jdk.internal.loader.ClassLoaders$AppClassLoader@67424e82'.
2020-04-29 22:09:37.084  INFO 24632 --- [           main] c.v.f.s.VaadinServletContextInitializer  : Search for VaadinAppShell took 30 ms
2020-04-29 22:09:37.085  INFO 24632 --- [           main] c.v.f.s.s.VaadinAppShellInitializer      : Using org.vaadin.example.AppShell class for configuring `index.html` response
HOTSWAP AGENT: 22:09:37.182 ERROR (org.hotswap.agent.command.ReflectionCommand) - Method servletInitialized not found in class org.hotswap.agent.plugin.vaadin.VaadinIntegration
java.lang.NoSuchMethodException: org.hotswap.agent.plugin.vaadin.VaadinIntegration.servletInitialized(com.vaadin.flow.spring.SpringServlet)
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2475)
	at org.hotswap.agent.command.ReflectionCommand.doExecuteReflectionCommand(ReflectionCommand.java:205)
	at org.hotswap.agent.command.ReflectionCommand.executeCommand(ReflectionCommand.java:168)
	at org.hotswap.agent.command.impl.CommandExecutor.run(CommandExecutor.java:43)

2020-04-29 22:09:38.986  INFO 24632 --- [           main] c.v.f.s.VaadinServletContextInitializer  : Search for subclasses and classes with annotations took 1850 ms
2020-04-29 22:09:38.987  INFO 24632 --- [           main] c.v.f.server.startup.DevModeInitializer  : Starting dev-mode updaters in /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring folder.
2020-04-29 22:09:39.008  INFO 24632 --- [           main] c.v.f.s.f.s.FullDependenciesScanner      : Visited 94 classes. Took 13 ms.
2020-04-29 22:09:39.020  INFO 24632 --- [nPool-worker-19] c.v.f.s.f.TaskGeneratePackageJson        : writing file /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/flow-frontend/package.json.
2020-04-29 22:09:39.027  INFO 24632 --- [nPool-worker-19] c.v.f.s.c.g.OpenApiSpecGenerator         : Parsing java files from [/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/src/main/java]
2020-04-29 22:09:39.181  INFO 24632 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'applicationTaskExecutor'
2020-04-29 22:09:39.205  INFO 24632 --- [nPool-worker-19] c.v.f.s.c.g.OpenApiSpecGenerator         : There are no connect endpoints to genertate.
2020-04-29 22:09:39.238  INFO 24632 --- [nPool-worker-19] c.v.f.s.frontend.TaskUpdatePackages      : Skipping `npm install`.
2020-04-29 22:09:39.238  INFO 24632 --- [nPool-worker-19] c.v.f.s.frontend.TaskCopyFrontendFiles   : Copying frontend resources from jar files ...
2020-04-29 22:09:39.290  INFO 24632 --- [nPool-worker-19] c.v.f.s.frontend.TaskCopyFrontendFiles   : Visited 13 resources. Took 51 ms.
2020-04-29 22:09:39.314  WARN 24632 --- [           main] o.s.b.a.m.MustacheAutoConfiguration      : Cannot find template location: classpath:/templates/ (please add some templates, check your Mustache configuration, or set spring.mustache.check-template-location=false)
2020-04-29 22:09:39.315  INFO 24632 --- [nPool-worker-19] c.v.f.server.frontend.TaskUpdateImports  : No js modules to update '/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/frontend/generated-flow-imports.js' file
2020-04-29 22:09:39.316  INFO 24632 --- [nPool-worker-19] c.v.f.server.frontend.TaskUpdateImports  : No js modules to update '/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/frontend/generated-flow-imports.d.ts' file
2020-04-29 22:09:39.317  INFO 24632 --- [onPool-worker-5] com.vaadin.flow.server.DevModeHandler    : Starting webpack-dev-server
2020-04-29 22:09:39.506  INFO 24632 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2020-04-29 22:09:39.512  INFO 24632 --- [           main] org.vaadin.example.Application           : Started Application in 4.263 seconds (JVM running for 4.902)

------------------ Starting Frontend compilation. ------------------
2020-04-29 22:09:39.600  INFO 24632 --- [onPool-worker-5] com.vaadin.flow.server.DevModeHandler    : Running webpack to compile frontend resources. This may take a moment, please stand by...
[0%] Compile modules...
[10%] Build modules...(0 of 0 modules )
ℹ ｢wds｣: Project is running at http://localhost:51015/webpack-dev-server/
ℹ ｢wds｣: webpack output is served from /VAADIN/
ℹ ｢wds｣: Content not from webpack is served from /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/target/classes/META-INF/VAADIN, src/main/webapp
[10%] Build modules...(0 of 1 modules frontend/index.ts)
ℹ ｢atl｣: Using typescript@3.8.3 from typescript
ℹ ｢atl｣: Using tsconfig.json from /Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/tsconfig.json

<snipped webpack progress output>

ℹ ｢wdm｣: Hash: 90081d924f32c9a6c5a8
Version: webpack 4.42.0
Time: 4613ms
Built at: 04/29/2020 10:09:45 PM
                                            Asset       Size  Chunks                                Chunk Names
     build/vaadin-0-c514abe5565e0451d1a4.cache.js     13 MiB       0  [emitted] [immutable]  [big]  
     build/vaadin-1-e1a564bc0e1712e27fc3.cache.js    319 KiB       1  [emitted] [immutable]         
     build/vaadin-2-4a4eff32f33c1a978fbe.cache.js    341 KiB       2  [emitted] [immutable]         
     build/vaadin-3-0f8424cb7ffacc6b2407.cache.js   21.7 KiB       3  [emitted] [immutable]         
     build/vaadin-4-c8dd7c2f4f5ac838b53f.cache.js    336 KiB       4  [emitted] [immutable]         
     build/vaadin-5-4129cd40035a2da764fc.cache.js   67.5 KiB       5  [emitted] [immutable]         
build/vaadin-bundle-c797b7cacdeb4785934a.cache.js    284 KiB  bundle  [emitted] [immutable]         bundle
                                       index.html  555 bytes          [emitted]                     
Entrypoint bundle = build/vaadin-bundle-c797b7cacdeb4785934a.cache.js
[../node_modules/@polymer/iron-icon/iron-icon.js] 5.5 KiB {0} [built]
[../node_modules/@polymer/iron-list/iron-list.js] 58.3 KiB {0} [built]
[../node_modules/@vaadin/router/dist/vaadin-router.js] 91 KiB {bundle} [built]
[../node_modules/@vaadin/vaadin-accordion/theme/lumo/vaadin-accordion.js] 78 bytes {0} [built]
[../node_modules/@vaadin/vaadin-app-layout/theme/lumo/vaadin-app-layout.js] 81 bytes {0} [built]
[../node_modules/@vaadin/vaadin-app-layout/theme/lumo/vaadin-drawer-toggle.js] 87 bytes {0} [built]
[../node_modules/@vaadin/vaadin-board/vaadin-board-row.js] 43 bytes {0} [built]
[../node_modules/@vaadin/vaadin-board/vaadin-board.js] 39 bytes {0} [built]
[../node_modules/@vaadin/vaadin-button/theme/lumo/vaadin-button.js] 73 bytes {0} [built]
[../node_modules/@vaadin/vaadin-charts/vaadin-chart.js] 74 bytes {0} [built]
[../node_modules/@vaadin/vaadin-checkbox/theme/lumo/vaadin-checkbox-group.js] 127 bytes {0} [built]
[../node_modules/@vaadin/vaadin-checkbox/theme/lumo/vaadin-checkbox.js] 77 bytes {0} [built]
[../target/flow-frontend/Flow.ts] 12.7 KiB {bundle} [built]
[../target/frontend/generated-flow-imports.js] 5.22 KiB {1} [built]
[./index.ts] 413 bytes {bundle} [built]
    + 474 hidden modules
Child html-webpack-plugin for "index.html":
     1 asset
    Entrypoint undefined = index.html
    [../node_modules/html-webpack-plugin/lib/loader.js!./index.html] 662 bytes {0} [built]
    [../node_modules/lodash/lodash.js] 528 KiB {0} [built]
    [../node_modules/webpack/buildin/global.js] 472 bytes {0} [built]
    [../node_modules/webpack/buildin/module.js] 497 bytes {0} [built]
ℹ ｢wdm｣: Compiled successfully.

----------------- Frontend compiled successfully. -----------------

2020-04-29 22:09:45.240  INFO 24632 --- [onPool-worker-5] com.vaadin.flow.server.DevModeHandler    : Started webpack-dev-server. Time: 5922ms
2020-04-29 22:09:54.207  INFO 24632 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2020-04-29 22:09:54.207  INFO 24632 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2020-04-29 22:09:54.213  INFO 24632 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 6 ms
2020-04-29 22:09:54.233  INFO 24632 --- [nio-8080-exec-2] c.v.f.s.DefaultDeploymentConfiguration   : 
=================== Vaadin DeploymentConfiguration ====================

 Vaadin is running in DEBUG MODE.
 When deploying application for production, remember to disable debug features. See more from https://vaadin.com/docs/

 Using Vaadin 15+ bootstrap mode.
 Using 'index.html' from '/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/frontend/index.html'
 Using 'index.ts' from '/Users/joheriks/vaadin/dev/skeleton-starter-flow-spring/frontend/index.ts'

=======================================================================
2020-04-29 22:09:55.008  INFO 24632 --- [nio-8080-exec-2] c.vaadin.flow.spring.SpringInstantiator  : The number of beans implementing 'I18NProvider' is 0. Cannot use Spring beans for I18N, falling back to the default behavior
HOTSWAP AGENT: 22:10:09.208 ERROR (org.hotswap.agent.command.ReflectionCommand) - Error executin method updateRoutes in class org.hotswap.agent.plugin.vaadin.VaadinIntegration
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.hotswap.agent.command.ReflectionCommand.doExecuteReflectionCommand(ReflectionCommand.java:207)
	at org.hotswap.agent.command.ReflectionCommand.executeCommand(ReflectionCommand.java:168)
	at org.hotswap.agent.plugin.vaadin.VaadinPlugin$UpdateRoutesCommand.executeCommand(VaadinPlugin.java:137)
	at org.hotswap.agent.command.impl.CommandExecutor.run(CommandExecutor.java:43)
Caused by: java.lang.NullPointerException
	at org.hotswap.agent.plugin.vaadin.VaadinIntegration.updateRoutes(VaadinIntegration.java:65)
	... 8 more

HOTSWAP AGENT: 22:10:10.858 ERROR (org.hotswap.agent.command.ReflectionCommand) - Error executin method reload in class org.hotswap.agent.plugin.vaadin.VaadinIntegration
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.hotswap.agent.command.ReflectionCommand.doExecuteReflectionCommand(ReflectionCommand.java:207)
	at org.hotswap.agent.command.ReflectionCommand.executeCommand(ReflectionCommand.java:168)
	at org.hotswap.agent.command.impl.CommandExecutor.run(CommandExecutor.java:43)
Caused by: java.lang.NullPointerException
	at org.hotswap.agent.plugin.vaadin.VaadinIntegration.reload(VaadinIntegration.java:93)
	... 7 more

```

--- a/flow-client/src/main/resources/META-INF/resources/frontend/Flow.ts
+++ b/flow-client/src/main/resources/META-INF/resources/frontend/Flow.ts
@@ -13,6 +13,7 @@ interface AppConfig {
   serviceUrl: string;
   springBootDevToolsPort: number;
   liveReloadBackend: string;
+  liveReloadPath: string;
 }
 
 interface AppInitResponse {
@@ -235,7 +236,7 @@ export class Flow {
       // (server ensures this parameter is true only in dev mode)
       if (appConfig.devmodeGizmoEnabled) {
         const devmodeGizmoMod = await import('./VaadinDevmodeGizmo');
-        const devmodeGizmo = await devmodeGizmoMod.init(appConfig.serviceUrl, appConfig.liveReloadBackend, appConfig.springBootDevToolsPort);
+        const devmodeGizmo = await devmodeGizmoMod.init(appConfig.serviceUrl, appConfig.liveReloadPath, appConfig.liveReloadBackend, appConfig.springBootDevToolsPort);
         $wnd.Vaadin.Flow.devModeGizmo = devmodeGizmo;
       }
 

--- a/flow-client/src/main/resources/META-INF/resources/frontend/VaadinDevmodeGizmo.d.ts
+++ b/flow-client/src/main/resources/META-INF/resources/frontend/VaadinDevmodeGizmo.d.ts
@@ -1,1 +1,1 @@
-export const init: (serviceUrl: string, liveReloadBackend: string, springBootDevToolsPort: number) => HTMLElement;
+export const init: (serviceUrl: string, liveReloadPath: string, liveReloadBackend: string, springBootDevToolsPort: number) => HTMLElement;

--- a/flow-client/src/main/resources/META-INF/resources/frontend/VaadinDevmodeGizmo.js
+++ b/flow-client/src/main/resources/META-INF/resources/frontend/VaadinDevmodeGizmo.js
@@ -402,6 +402,7 @@ class VaadinDevmodeGizmo extends LitElement {
       notifications: {type: Array},
       status: {type: String},
       serviceurl: {type: String},
+      liveReloadPath: {type: String},
       liveReloadBackend: {type: String},
       springBootDevToolsPort: {type: Number}
     };
@@ -534,19 +535,32 @@ class VaadinDevmodeGizmo extends LitElement {
   }
 
   openDedicatedWebSocketConnection() {
-    const url = this.serviceurl ? this.serviceurl : window.location.toString();
+    const wsUrl = this.getDedicatedWebSocketUrl(window.location);
+    if (wsUrl) {
+      this.connection = new WebSocket(wsUrl);
+      const self = this;
+      setInterval(function() {
+        if (self.connection !== null) {
+          self.connection.send('');
+        }
+      }, VaadinDevmodeGizmo.HEARTBEAT_INTERVAL);
+    }
+  }
+
+  getDedicatedWebSocketUrl(location) {
+    let url;
+    if (this.serviceurl) {
+      url = this.serviceurl;
+    } else if (this.liveReloadPath) {
+      url = location.protocol + '//' + location.host + '/' + this.liveReloadPath;
+    } else {
+      url = location.href;
+    }
     if (!url.startsWith('http://') && !url.startsWith('https://')) {
       console.warn('The protocol of the url should be http or https for live reload to work.');
-      return;
+      return null;
     }
-    const wsUrl = url.replace(/^http/, 'ws') + '?refresh_connection';
-    const self = this;
-    this.connection = new WebSocket(wsUrl);
-    setInterval(function() {
-      if (self.connection !== null) {
-        self.connection.send('');
-      }
-    }, VaadinDevmodeGizmo.HEARTBEAT_INTERVAL);
+    return url.replace(/^http/, 'ws') + '?v-r=push&refresh_connection';
   }
 
   getSpringBootWebSocketUrl(location) {
@@ -817,12 +831,17 @@ class VaadinDevmodeGizmo extends LitElement {
   }
 }
 
-const init = function(serviceUrl, liveReloadBackend, springBootDevToolsPort) {
+const init = function(serviceUrl, liveReloadPath, liveReloadBackend, springBootDevToolsPort) {
   if ('false' !== window.localStorage.getItem(VaadinDevmodeGizmo.ENABLED_KEY_IN_LOCAL_STORAGE)) {
-    customElements.define('vaadin-devmode-gizmo', VaadinDevmodeGizmo);
+    if (customElements.get('vaadin-devmode-gizmo') === undefined) {
+      customElements.define('vaadin-devmode-gizmo', VaadinDevmodeGizmo);
+    }
     const devmodeGizmo = document.createElement('vaadin-devmode-gizmo');
     if (serviceUrl) {
       devmodeGizmo.setAttribute('serviceurl', serviceUrl);
+    }
+    if (liveReloadPath) {
+      devmodeGizmo.setAttribute('liveReloadPath', liveReloadPath);
     }
     if (liveReloadBackend) {
       devmodeGizmo.setAttribute('liveReloadBackend', liveReloadBackend);

--- a/flow-client/src/test/frontend/VaadinDevmodeGizmoTests.js
+++ b/flow-client/src/test/frontend/VaadinDevmodeGizmoTests.js
@@ -6,12 +6,23 @@ import { init } from "../../main/resources/META-INF/resources/frontend/VaadinDev
 describe('VaadinDevmodeGizmo', () => {
 
   it('should connect to port-hostname.gitpod.io with Spring Boot Devtools', () => {
-    let gizmo = init('', 'SPRING_BOOT_DEVTOOLS', 35729);
+    let gizmo = init(undefined, '','SPRING_BOOT_DEVTOOLS', 35729);
     let location = {
       'protocol': 'https',
       'hostname': 'abc-12345678-1234-1234-1234-1234567890ab.ws-eu01.gitpod.io'
     };
     assert.equal(gizmo.getSpringBootWebSocketUrl(location),
       'ws://35729-12345678-1234-1234-1234-1234567890ab.ws-eu01.gitpod.io');
+  });
+
+  it('should append push target if given', () => {
+    let gizmo = init(undefined, 'context/vaadinServlet','HOTSWAP_AGENT', 35729);
+    let location = {
+      'href': 'http://localhost:8080/context/myroute',
+      'protocol': 'http:',
+      'host': 'localhost:8080',
+    };
+    assert.equal(gizmo.getDedicatedWebSocketUrl(location),
+      'ws://localhost:8080/context/vaadinServlet?v-r=push&refresh_connection');
   });
 });

--- a/flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java
@@ -1054,7 +1054,7 @@ public class BootstrapHandler extends SynchronizedRequestHandler {
          * @return A non-null {@link JsonObject} with application parameters.
          */
         public JsonObject getApplicationParameters(BootstrapContext context) {
-            JsonObject appConfig = getApplicationParameters(
+            JsonObject appConfig = getApplicationParameters(context,
                     context.getRequest(), context.getSession());
 
             appConfig.put(ApplicationConstants.UI_ID_PARAMETER,
@@ -1062,7 +1062,7 @@ public class BootstrapHandler extends SynchronizedRequestHandler {
             return appConfig;
         }
 
-        private JsonObject getApplicationParameters(VaadinRequest request,
+        private JsonObject getApplicationParameters(BootstrapContext context, VaadinRequest request,
                 VaadinSession session) {
             DeploymentConfiguration deploymentConfiguration = session
                     .getConfiguration();
@@ -1093,6 +1093,11 @@ public class BootstrapHandler extends SynchronizedRequestHandler {
                 if (liveReload != null && liveReload.getBackend() != null) {
                     appConfig.put("liveReloadBackend",
                             liveReload.getBackend().toString());
+                    String pushURL = session.getConfiguration().getPushURL();
+                    if (pushURL != null) {
+                        appConfig.put("liveReloadPath", context
+                                .getUriResolver().resolveVaadinUri(pushURL));
+                    }
                 }
 
                 // make configurable when fixing #7847

